docs/configuration.rst
Configuration
=============

/etc/graphite-api.yaml
----------------------

The configuration file for Graphite-API lives at ``/etc/graphite-api.yaml``
and uses the YAML format. Creating the configuration file is optional: if
Graphite-API doesn't find the file, sane defaults are used. They are described
below.

Default values
``````````````

.. code-block:: yaml

    search_index: /srv/graphite/index
    finders:
      - graphite_api.finders.whisper.WhisperFinder
    functions:
      - graphite_api.functions.SeriesFunctions
      - graphite_api.functions.PieFunctions
    whisper:
      directories:
        - /srv/graphite/whisper
    time_zone: <system timezone> or UTC

Config sections
```````````````

Default sections
^^^^^^^^^^^^^^^^

*search_index*

  The location of the search index used for searching metrics. Note that it
  needs to be a file that is writable by the Graphite-API process.

*finders*

  A list of python paths to the storage finders you want to use when fetching
  metrics.

*functions*

  A list of python paths to function definitions for transforming / analyzing
  time series data.

*whisper*

  The configuration information for whisper. Only relevant when using
  WhisperFinder. Simply holds a ``directories`` key listing all directories
  containing whisper data.

*time_zone*

  The time zone to use when generating graphs. By default, Graphite-API tries
  to detect your system timezone. If detection fails it falls back to UTC. You
  can also manually override it if you want another value than your system's
  timezone.

Extra sections
^^^^^^^^^^^^^^

*carbon*

  Configuration information for reading data from carbon's cache. Items:

  *hosts*
    List of carbon-cache hosts, in the format ``hostname:port[:instance]``.

  *timeout*
    Socket timeout for carbon connections, in seconds.

  *retry_delay*
    Time to wait before trying to re-establish a failed carbon connection, in
    seconds.

  *hashing_keyfunc*
    Python path to a hashing function for metrics. If you use Carbon with
    consistent hashing and a custom function, you need to point to the same
    hashing function.

  *hashing_type*
    Type of metric hashing function. The default ``carbon_ch`` is Graphite's
    traditional consistent-hashing implementation. Alternatively, you can use
    ``fnv1a_ch``, which supports the Fowler-Noll-Vo hash function (FNV-1a) hash
    implementation offered by the carbon-c-relay project.
    Default: ``carbon_ch``

  *carbon_prefix*
    Prefix for carbon's internal metrics. When querying metrics starting with
    this prefix, requests are made to all carbon-cache instances instead of
    one instance selected by the key function. Default: ``carbon``.

  *replication_factor*
     The replication factor of your carbon setup. Default: ``1``.

  Example:

  .. code-block:: yaml

      carbon:
        hosts:
          - 127.0.0.1:7002
        timeout: 1
        retry_delay: 15
        carbon_prefix: carbon
        replication_factor: 1

*sentry_dsn*

  This is useful if you want to send Graphite-API's exceptions to a `Sentry`_
  instance for easier debugging.

  Example::

      sentry_dsn: https://key:secret@app.getsentry.com/12345

  .. note::

      Sentry integration requires Graphite-API to be installed with the
      corresponding extra dependency::

          $ pip install graphite-api[sentry]

.. _Sentry: https://docs.getsentry.com

*allowed_origins*

  Allows you to do cross-domain (CORS) requests to the Graphite API. Say you
  have a dashboard at ``dashboard.example.com`` that makes AJAX requests to
  ``graphite.example.com``, just set the value accordingly::

      allowed_origins:
        - dashboard.example.com

  You can specify as many origins as you want. A wildcard can be used to allow
  all origins::

      allowed_origins:
        - *

*cache*

  Lets you configure a cache for graph rendering. This is done via
  `Flask-Cache <http://pythonhosted.org/Flask-Cache/>`_ which supports a
  number of backends including memcache, Redis, filesystem or in-memory
  caching.

  Cache configuration maps directly to Flask-Cache's config values. For each
  ``CACHE_*`` config value, set the lowercased name in the ``cache`` section,
  without the prefix. Example::

      cache:
        type: redis
        redis_host: localhost

  This would configure Flask-Cache with ``CACHE_TYPE = 'redis'`` and
  ``CACHE_REDIS_HOST = 'localhost'``.

  Some cache options have default values defined by Graphite-API:

  * ``default_timeout``: 60

  * ``key_prefix``: ``'graphite-api:``.

  .. note::

      Caching functionality requires you to install the cache extra dependency
      but also the underlying driver. E.g. for redis, you'll need::

          $ pip install graphite-api[cache] redis

*statsd*

  Attaches a statsd object to the application, which can be used for
  instrumentation. Currently Graphite-API itself doesn't use this,
  but some backends do, like `Graphite-Influxdb`_.

  Example::

      statsd:
          host: 'statsd_host'
          port: 8125  # not needed if default

  .. note::

      This requires the statsd module::

          $ pip install statsd

.. _Graphite-Influxdb: https://github.com/vimeo/graphite-influxdb

*render_errors*

  If ``True`` (default), full tracebacks are returned in the HTTP
  response in case of application errors.

Custom location
---------------

If you need the Graphite-API config file to be stored in another place than
``/etc/graphite-api.yaml``, you can set a custom location using the
``GRAPHITE_API_CONFIG`` environment variable::

    export GRAPHITE_API_CONFIG=/var/lib/graphite/config.yaml


docs/releases.rst
Graphite-API releases
=====================

1.1.3 -- 2016-05-23
-------------------

* Remove extra parenthesis from ``aliasByMetric()``.
* Fix leap year handling in ``graphite_api.render.attime``.
* Allow colon and hash in node names in ``aliasByNode()``
* Fix calling ``reduceFunction`` in ``reduceSeries``
* Revert a whisper patch which broke multiple retentions handling.
* Specify which function is invalid when providing an invalid consolidation
  function.

1.1.2 -- 2015-11-19
-------------------

* Fix regression in multi fetch handling: paths were queried multiple times,
  leading to erroneous behaviour and slowdown.
* Continue on IndexError in ``remove{Above,Below}Percentile`` functions.

1.1.1 -- 2015-10-23
-------------------

* Fix ``areaMode=stacked``.

* Fix error when calling functions that use ``fetchWithBootstrap`` and the
  bootstrap range isn't available (fill with nulls instead).

1.1 -- 2015-10-05
-----------------

* Add CarbonLink support.

* Add support for configuring a cache backend and the ``noCache`` and
  ``cacheTimeout`` API options.

* When no timezone is provided in the configuration file, try to guess from
  the system's timezone with a fallback to UTC.

* Now supporting Flask >= 0.8 and Pyparsing >= 1.5.7.

* Add support for ``fetch_multi()`` in storage finders. This is useful for
  database-backed finders such as Cyanite because it allows fetching all time
  series at once instead of sequentially.

* Add ``multiplySeriesWithWildcards``, ``minimumBelow``, ``changed``,
  ``timeSlice`` and ``removeEmptySeries`` functions.

* Add optional ``step`` argument to ``time``, ``sin`` and ``randomWalk``
  functions.

* Add ``/metrics`` API call as an alias to ``/metrics/find``.

* Add missing ``/metrics/index.json`` API call.

* Allow wildcards origins (``*``) in CORS configuration.

* Whisper finder now logs debug information.

* Fix parsing dates such as "feb27" during month days > 28.

* Change ``sum()`` to return ``null`` instead of 0 when all series' datapoints
  are null at the same time. This is graphite-web's behavior.

* Extract paths of all targets before fetching data. This is a significant
  optimization for storage backends such as Cyanite that allow bulk-fetching
  metrics.

* Add JSONP support to all API endpoints that can return JSON.

* Fix 500 error when generating a SVG graph without any data.

* Return tracebacks in the HTTP response when app errors occur. This behavior
  can be disabled in the configuration.

* Fixes for the following graphite-web issues:

  * `#639 <https://github.com/graphite-project/graphite-web/issues/639>`_ --
    proper timezone handling of ``from`` and ``until`` with client-supplied
    timezones.
  * `#540 <https://github.com/graphite-project/graphite-web/issues/540>`_ --
    provide the last data point when rendering to JSON format.
  * `#381 <https://github.com/graphite-project/graphite-web/issues/381>`_ --
    make ``areaBetween()`` work either when passed 2 arguments or a single
    wildcard series of length 2.
  * `#702 <https://github.com/graphite-project/graphite-web/pull/702>`_ --
    handle backslash as path separator on windows.
  * `#410 <https://github.com/graphite-project/graphite-web/pull/410>`_ -- SVG
    output sometimes had an extra ``</g>`` tag.

1.0.1 -- 2014-03-21
-------------------

* ``time_zone`` set to UTC by default instead of Europe/Berlin.
* Properly log app exceptions.
* Fix constantLine for python 3.
* Create whisper directories if they don't exist.
* Fixes for the following graphite-web issues:

  * `#645 <https://github.com/graphite-project/graphite-web/pull/645>`_, `#625
    <https://github.com/graphite-project/graphite-web/issues/625>`_ -- allow
    ``constantLine`` to work even if there are no other targets in the graph.

1.0.0 -- 2014-03-20
-------------------

Version 1.0 is based on the master branch of Graphite-web, mid-March 2014,
with the following modifications:

* New ``/index`` API endpoint for re-building the index (replaces the
  build-index command-line script from graphite-web).

* Removal of memcache integration.

* Removal of Pickle integration.

* Removal of remote rendering.

* Support for Python 3.

* A lot more tests and test coverage.

* Fixes for the following graphite-web issues:

  * (meta) `#647 <https://github.com/graphite-project/graphite-web/issues/647>`_
    -- strip out the API from graphite-web.
  * `#665 <https://github.com/graphite-project/graphite-web/pull/665>`_ --
    address some DeprecationWarnings.
  * `#658 <https://github.com/graphite-project/graphite-web/issues/658>`_ --
    accept a float value in ``maxDataPoints``.
  * `#654 <https://github.com/graphite-project/graphite-web/pull/654>`_ --
    ignore invalid ``logBase`` values (<=1).
  * `#591 <https://github.com/graphite-project/graphite-web/issues/591>`_ --
    accept JSON data additionaly to querystring params or form data.


graphite_api/config.py
import logging
import os
import traceback
import warnings
from importlib import import_module

import structlog
import yaml
from flask import make_response
from structlog.processors import (format_exc_info, JSONRenderer,
                                  KeyValueRenderer)
from tzlocal import get_localzone

from . import DEBUG
from .middleware import CORS, TrailingSlash
from .storage import Store

try:
    from logging.config import dictConfig
except ImportError:
    from logutils.dictconfig import dictConfig

if DEBUG:
    processors = (format_exc_info, KeyValueRenderer())
else:
    processors = (format_exc_info, JSONRenderer())

logger = structlog.get_logger()

default_conf = {
    'search_index': '/srv/graphite/index',
    'finders': [
        'graphite_api.finders.whisper.WhisperFinder',
    ],
    'functions': [
        'graphite_api.functions.SeriesFunctions',
        'graphite_api.functions.PieFunctions',
    ],
    'whisper': {
        'directories': [
            '/srv/graphite/whisper',
        ],
    },
    'time_zone': get_localzone().zone,
}
if default_conf['time_zone'] == 'local':  # tzlocal didn't find anything
    default_conf['time_zone'] = 'UTC'


# attributes of a classical log record
NON_EXTRA = set(['module', 'filename', 'levelno', 'exc_text', 'pathname',
                 'lineno', 'msg', 'funcName', 'relativeCreated',
                 'levelname', 'msecs', 'threadName', 'name', 'created',
                 'process', 'processName', 'thread'])


class StructlogFormatter(logging.Formatter):
    def __init__(self, *args, **kwargs):
        self._bound = structlog.BoundLoggerBase(None, processors, {})

    def format(self, record):
        if not record.name.startswith('graphite_api'):
            kw = dict(((k, v) for k, v in record.__dict__.items()
                       if k not in NON_EXTRA))
            kw['logger'] = record.name
            return self._bound._process_event(
                record.levelname.lower(), record.getMessage(), kw)[0]
        return record.getMessage()


def load_by_path(path):
    module, klass = path.rsplit('.', 1)
    finder = import_module(module)
    return getattr(finder, klass)


def error_handler(e):
    return make_response(traceback.format_exc(), 500,
                         {'Content-Type': 'text/plain'})


def configure(app):
    config_file = os.environ.get('GRAPHITE_API_CONFIG',
                                 '/etc/graphite-api.yaml')
    if os.path.exists(config_file):
        with open(config_file) as f:
            config = yaml.safe_load(f)
            config['path'] = config_file
    else:
        warnings.warn("Unable to find configuration file at {0}, using "
                      "default config.".format(config_file))
        config = {}

    configure_logging(config)

    for key, value in list(default_conf.items()):
        config.setdefault(key, value)

    app.statsd = None
    if 'statsd' in config:
        try:
            from statsd import StatsClient
        except ImportError:
            warnings.warn("'statsd' is provided in the configuration but "
                          "the statsd client is not installed. Please `pip "
                          "install statsd`.")
        else:
            c = config['statsd']
            app.statsd = StatsClient(c['host'], c.get('port', 8125))

    app.cache = None
    if 'cache' in config:
        try:
            from flask.ext.cache import Cache
        except ImportError:
            warnings.warn("'cache' is provided in the configuration but "
                          "Flask-Cache is not installed. Please `pip install "
                          "Flask-Cache`.")
        else:
            cache_conf = {'CACHE_DEFAULT_TIMEOUT': 60,
                          'CACHE_KEY_PREFIX': 'graphite-api:'}
            for key, value in config['cache'].items():
                cache_conf['CACHE_{0}'.format(key.upper())] = value
            app.cache = Cache(app, config=cache_conf)

    loaded_config = {'functions': {}}
    for functions in config['functions']:
        loaded_config['functions'].update(load_by_path(functions))

    if 'carbon' in config:
        if 'hashing_keyfunc' in config['carbon']:
            config['carbon']['hashing_keyfunc'] = load_by_path(
                config['carbon']['hashing_keyfunc'])
        else:
            config['carbon']['hashing_keyfunc'] = lambda x: x
    loaded_config['carbon'] = config.get('carbon', None)

    finders = []
    for finder in config['finders']:
        finders.append(load_by_path(finder)(config))
    loaded_config['store'] = Store(finders)
    app.config['GRAPHITE'] = loaded_config
    app.config['TIME_ZONE'] = config['time_zone']
    logger.info("configured timezone", timezone=app.config['TIME_ZONE'])

    if 'sentry_dsn' in config:
        try:
            from raven.contrib.flask import Sentry
        except ImportError:
            warnings.warn("'sentry_dsn' is provided in the configuration but "
                          "the sentry client is not installed. Please `pip "
                          "install raven[flask]`.")
        else:
            Sentry(app, dsn=config['sentry_dsn'])

    app.wsgi_app = TrailingSlash(CORS(app.wsgi_app,
                                      config.get('allowed_origins')))
    if config.get('render_errors', True):
        app.errorhandler(500)(error_handler)


def configure_logging(config):
    structlog.configure(processors=processors,
                        logger_factory=structlog.stdlib.LoggerFactory(),
                        wrapper_class=structlog.stdlib.BoundLogger,
                        cache_logger_on_first_use=True)
    config.setdefault('logging', {})
    config['logging'].setdefault('version', 1)
    config['logging'].setdefault('handlers', {})
    config['logging'].setdefault('formatters', {})
    config['logging'].setdefault('loggers', {})
    config['logging']['handlers'].setdefault('raw', {
        'level': 'DEBUG',
        'class': 'logging.StreamHandler',
        'formatter': 'raw',
    })
    config['logging']['loggers'].setdefault('root', {
        'handlers': ['raw'],
        'level': 'DEBUG',
        'propagate': False,
    })
    config['logging']['loggers'].setdefault('graphite_api', {
        'handlers': ['raw'],
        'level': 'DEBUG',
    })
    config['logging']['formatters']['raw'] = {'()': StructlogFormatter}
    dictConfig(config['logging'])
    if 'path' in config:
        logger.info("loading configuration", path=config['path'])
    else:
        logger.info("loading default configuration")
